test/inference-engines/helpers.ts

```diff
@@ -0,0 +1,184 @@
+
+/**
+ * Round a percentage to the desired precision.
+ * @param x The number to be rounded
+ */
+export const roundResult = (x: number) => {
+  if (x === 0) return x
+  const scale = Math.pow(10, 8 + Math.floor(-Math.log10(x)))
+  return Math.round(x * scale) / scale
+}
+
+/**
+ * The test case type represents a test case case
+ * to be run.  The event and evidence fields provide the
+ * arguments to the probabilistic inference to compute
+ * and the expected field contains the expected value.
+ */
+export type TestCase = {
+  event: Record<string, string[]>;
+  evidence: Record<string, string[]>;
+  expected: number;
+}
+
+/**
+ * This algorithm provides a substantially different means of computing
+ * a probabilistic inference by filterning and summing the rows in a
+ * gold standard dataset.
+ * @param names A list of the names of the varialbes in the dataset
+ * @param GOLD_STANDARD An independently verified dataset containing
+ *   the joint probability of every possible combination of variable's
+ *   values.
+ * @returns a function that when given the event will give you the
+ *   probability by summing over the rows in the dataset.
+ */
+export const aggregateGoldStandardResults = (names: string[], GOLD_STANDARD: [string[], number][]) => (event: Record<string, string[]>): number => {
+  // filter the dataset to find all the rows that are not in conflict with
+  // the evidence.
+  const rows = GOLD_STANDARD.filter(kv => names.every((name, i) => {
+    const vs = event[name]
+    return vs == null || vs.includes(kv[0][i])
+  }))
+  // sum over the rows to compute the probability.
+  return sum(rows.map(kv => kv[1]))
+}
+
+/**
+ * This helper function constructs all the distinct possible events that can
+ * be constructed by choosing one of the provided test values for each
+ * variable.   The probability of each event is computed using the gold
+ * standard dataset.
+ * @param names the names of the variables in the dataset
+ * @param testValues A list of possible values for each of the variables
+ * @param goldStandard An independently verified dataset containing
+ *   the joint probability of every possible combination of variable's
+ *   values.
+ * @returns a collection of test cases that can be passed to the test
+ *   runner.
+ */
+function makeTestCases (names: string[], testValues: string[][][], goldStandard: [string[], number][]): TestCase[] {
+  const testCases: { event: Record<string, string[]>; evidence: Record<string, string[]>; expected: number}[] = []
+  const numberOfTestCases = product(testValues.map(xs => xs.length * 3))
+  const aggregate = aggregateGoldStandardResults(names, goldStandard)
+
+  for (let i = 0; i < numberOfTestCases; i++) {
+    let varIdx = 0
+    let x = i
+    const event: Record<string, string[]> = {}
+    const evidenceAndEvent: Record<string, string[]> = {}
+    const evidence: Record<string, string[]> = {}
+    // for each variable pick a value and a location in either the
```

Comment:
nit: Sentence needs to start with a capital.

Reply:
fixed